README.md
# shelf_secure_cookie
Encrypted cookies use `base64Url` encoding. See `example/example.dart` for details.

Based on `shelf_cookie` package, added async `setEncrypted` and `getEncrypted` cookie methods, that support `AES-GCM` cipher with digest. These should be used to store sensitive data, if you prefer a cookie storage.

Both methods require a 32-byte secret key. You can use a key with a different
length, but be warned that it will be padded with zeroes and trimed to 32
bytes exactly.

This package is not battle-tested yet, but I'm working on this. Also if you happened to download 1.0.0 version, update to >=1.0.2 as soon as possible.

## Original docs

Cookie parser middleware for the Dart Shelf ecosystem.
Reads cookies in request, sets cookies in response.

Adds a `CookieParser` instance to `request.context['cookies']` to help
manipulate cookies.

## Example

```dart
import 'dart:io';
import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf_cookie/shelf_cookie.dart';

/// Handle a request that contains a `Cookie` header.
/// e.g. 'Cookie': 'ping=foo'
var handler = const shelf.Pipeline()
    // initialize cookie parser middleware
    .addMiddleware(cookieParser())
    .addHandler((req) async {
  CookieParser cookies = req.context['cookies'];

  // Retrieve request cookies.
  var reqCookie = cookies.get('ping');
  print(reqCookie.name); // foo

  // Clear cookies because Shelf currently only supports
  // a single `Set-Cookie` header in response.
  cookies.clear();

  // Create a cookie for response.
  var resCookie = cookies.set('pong', 'bar', secure: true);

  // Middleware will add `Set-Cookie` response header.
  // e.g. 'Set-Cookie': 'pong=bar; Secure; HttpOnly'
  return shelf.Response.ok('OK', headers: {HttpHeaders.setCookieHeader: cookies.toHeader()});
});
```

## TODO
Add handy Request & Response extensions or adapt `cookieParser()` middleware to read newly set cookies. The original version of middleware required to call `cookies.clear()` every time before setting new values and if you forget this, it became messy. So now you have to set headers explicitly, but hopefully not for long.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.3.0] - 2022-01-17
### Changed
- Allowed to use a key with different than 32-bytes length, what is more
  user-friendly, but somewhat less secure. The key is padded with zeroes and
  trimmed to 32 bytes automatically. Backward compatible.

## [1.2.0] - 2021-04-27
### Changed
- Changed original middleware, now you don't have to set response 'Set-Cookie' headers manually, it will handle it if you operate on Request.context['cookies'] storage, see example.
- Now CookieParser have separated lists of Request cookies [cookies] and Response cookies [responseCookies], so this is a somewhat a breaking change.
- Now 'get' methods operate on a request cookies list, and 'set' on a response list. The exception is a 'getResponse' method that allows to read a newly set cookie if you need that.

## [1.1.0] - 2021-04-27
### Fixed
- Changed original middleware, now you have to set response headers explicitly, see example. But I will rearrange this later.
- Now encrypted cookies use `base64Url` encoding instead of `base64`.

## [1.0.6] - 2021-04-27
### Fixed
- Fixed CHANGELOG

## [1.0.5] - 2021-04-27
### Fixed
- Fixed middleware to accept secureKey
- Updated example and readme

## [1.0.2] - 2021-04-27
### Fixed
- Changed API, now secureKey is set at construction, get/setEncrypted methods use the same signature as get/set

## [1.0.1] - 2021-04-27
### Fixed
- Fixed security issue

## [1.0.0] - 2021-04-10
### Added
- Cookie parser with setEncrypted & getEncrypted methods
- All original code copied from null-safe branch of shelf_cookie


test/cookie_parser_test.dart
import 'dart:io';
import 'dart:convert';

import 'package:cryptography/cryptography.dart';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import 'package:shelf_secure_cookie/shelf_secure_cookie.dart';
import 'package:test/test.dart';

class NoCookieException implements Exception {}

Matcher throwsNoCookieException = throwsA(isA<NoCookieException>());

class TestRouter {
  Handler get handler {
    final router = Router();

    router.get('/gets', (Request request) async {
      var cookies = request.context["cookies"] as CookieParser;
      if (cookies.get("user") == null) throw NoCookieException();
      return Response.ok("a");
    });
    router.get('/sets', (Request request) async {
      var cookies = request.context["cookies"] as CookieParser;
      cookies.set("user", "1");
      return Response.ok("a");
    });

    var handler =
        const Pipeline().addMiddleware(cookieParser()).addHandler(router);

    return handler;
  }
}

void main() {
  late final Handler router = TestRouter().handler;

  test('isEmpty is true if header is empty', () {
    var cookies = CookieParser.fromCookieValue(null);
    expect(cookies.isEmpty, isTrue);
  });

  test('isEmpty is true if default constructor used', () {
    var cookies = CookieParser();
    expect(cookies.isEmpty, isTrue);
  });

  test('isResponseEmpty is true if header is empty', () {
    var cookies = CookieParser.fromCookieValue(null);
    expect(cookies.isResponseEmpty, isTrue);
  });

  test('isResponseEmpty is true if default constructor used', () {
    var cookies = CookieParser();
    expect(cookies.isResponseEmpty, isTrue);
  });

  test('parses cookies from `cookie` header value', () {
    var cookies = CookieParser.fromCookieValue('foo=bar; baz=qux');
    expect(cookies.isEmpty, isFalse);
    expect(cookies.get('foo')!.value, equals('bar'));
    expect(cookies.get('baz')!.value, equals('qux'));
  });

  test('parses cookies from raw headers map', () {
    var cookies =
        CookieParser.fromHeader({HttpHeaders.cookieHeader: 'foo=bar; baz=qux'});
    expect(cookies.isEmpty, isFalse);
    expect(cookies.get('foo')!.value, equals('bar'));
    expect(cookies.get('baz')!.value, equals('qux'));
  });

  test('adds new cookie to responseCookies list', () {
    var cookies = CookieParser();
    expect(cookies.getResponse('baz'), isNull);
    cookies.set('baz', 'qux');
    expect(cookies.getResponse('baz')!.value, 'qux');
  });

  test('removes cookie from cookies list by name', () {
    var cookies = CookieParser();
    cookies.set('baz', 'qux');
    cookies.remove('baz');
    expect(cookies.getResponse('baz'), isNull);
  });

  test('clears all cookies in list', () {
    var cookies = CookieParser();
    cookies.set('baz', 'qux');
    cookies.clear();
    expect(cookies.isResponseEmpty, isTrue);
  });

  //encrypted cookies
  test('encodes and ciphers an encrypted cookie', () async {
    final keyStr = "12345678901234567890123456789012";
    var cookies = CookieParser(keyStr);
    await cookies.setEncrypted('baz', 'qux');
    var cookie = cookies.getResponse('baz');
    var decoded = base64.decode(cookie!.value);
    final algorithm = AesGcm.with256bits();
    final key = await algorithm.newSecretKeyFromBytes(utf8.encode(keyStr));
    //can't use fromConcatenation constructor because of this:
    //https://github.com/dint-dev/cryptography/issues/55
    //final box = SecretBox.fromConcatenation(decoded, nonceLength: 12, macLength: 16);
    final cipherText = decoded.skip(12).take(decoded.length - 12 - 16).toList();
    final nonce = decoded.take(12).toList();
    final mac =
        decoded.skip(nonce.length + cipherText.length).take(16).toList();
    final box = SecretBox(cipherText, nonce: nonce, mac: Mac(mac));
    final bytes = await algorithm.decrypt(box, secretKey: key);
    final value = utf8.decode(bytes);
    expect(value, 'qux');
  });

  test('decodes and deciphers an encrypted cookie', () async {
    final keyStr = "12345678901234567890123456789012";
    var cookies = CookieParser(keyStr);
    var cookie = await cookies.setEncrypted('baz', 'qux');
    //as cookies list is split now, need to create a new parser as if I'm working in a subsequent request
    var cookie2 = Cookie.fromSetCookieValue(cookie.toString());
    var cookies2 = CookieParser(keyStr)..cookies.add(cookie2);
    var cookie3 = await cookies2.getEncrypted('baz');
    expect(cookie3, isNotNull);
    expect(cookie3!.value, 'qux');
  });

  test('works with key less than 32 bytes in length', () async {
    final keyStr = "12345678901234567890";
    var cookies = CookieParser(keyStr);
    var cookie = await cookies.setEncrypted('baz', 'qux');
    //as cookies list is split now, need to create a new parser as if I'm working in a subsequent request
    var cookie2 = Cookie.fromSetCookieValue(cookie.toString());
    var cookies2 = CookieParser(keyStr)..cookies.add(cookie2);
    var cookie3 = await cookies2.getEncrypted('baz');
    expect(cookie3, isNotNull);
    expect(cookie3!.value, 'qux');
  });

  test('works with key more than 32 bytes in length', () async {
    final keyStr =
        "123456789012345678901234567890129999999999999999999999999999";
    var cookies = CookieParser(keyStr);
    var cookie = await cookies.setEncrypted('baz', 'qux');
    //as cookies list is split now, need to create a new parser as if I'm working in a subsequent request
    var cookie2 = Cookie.fromSetCookieValue(cookie.toString());
    var cookies2 = CookieParser(keyStr)..cookies.add(cookie2);
    var cookie3 = await cookies2.getEncrypted('baz');
    expect(cookie3, isNotNull);
    expect(cookie3!.value, 'qux');
  });

  test('Middleware reads cookies and does not write if not told to', () async {
    final cp = CookieParser();
    final cookie = cp.set("user", "1");
    final request = Request(
      'GET',
      Uri.parse('http://localhost:8080/gets'),
      headers: {"cookie": cookie.toString()},
    );

    expect(() async => await router(request), returnsNormally);

    final Response response = await router(request);
    expect(response.headers[HttpHeaders.setCookieHeader], null,
        reason: "Does not retranslate cookies as before");
  });

  test('Middleware throws NoCookieException if expected cookie not set',
      () async {
    final cp = CookieParser();
    final cookieWrong = cp.set("user2", "1");
    final request = Request(
      'GET',
      Uri.parse('http://localhost:8080/gets'),
      headers: {"cookie": cookieWrong.toString()},
    );
    expect(() async => await router(request), throwsNoCookieException);
  });

  test('Middleware sets a cookie', () async {
    final request = Request(
      'GET',
      Uri.parse('http://localhost:8080/sets'),
    );
    expect(() async => await router(request), returnsNormally);
    final response = await router(request);
    expect(response, isNotNull);
    expect(response.headers[HttpHeaders.setCookieHeader], isNotNull);
    final cookie = Cookie.fromSetCookieValue(
        response.headers[HttpHeaders.setCookieHeader]!);
    expect(cookie.name, "user");
    expect(cookie.value, "1");
  });
}


lib/src/shelf_cookie.dart
import 'dart:async';
import 'dart:io';
import 'package:shelf/shelf.dart';

import 'cookie_parser.dart';

/// Creates a Shelf [Middleware] to parse cookies.
///
/// Adds a [CookieParser] instance to `request.context['cookies']`,
/// with convenience methods to manipulate cookies in request handlers.
///
/// Adds a `Set-Cookie` HTTP header to the response with all cookies.

Middleware cookieParser([String secretKey = ""]) {
  return (Handler innerHandler) {
    return (Request request) {
      final cookies = CookieParser.fromHeader(request.headers, secretKey);
      return Future.sync(() {
        return innerHandler(
          request.change(context: {'cookies': cookies}),
        );
      }).then((Response response) {
        return cookies.isResponseEmpty
            ? response
            : response.change(
                headers: {HttpHeaders.setCookieHeader: cookies.toHeader()});
      });
    };
  };
}


lib/src/cookie_parser.dart
import 'dart:io';
import 'dart:convert';

import 'package:collection/collection.dart' show IterableExtension;
import 'package:cryptography/cryptography.dart';
import 'package:shelf/shelf.dart';

/// Parses cookies from the `Cookie` header of a [Request].
///
/// Stores all cookies in a [cookies] list, and has convenience
/// methods to manipulate this list.
class CookieParser {
  //There were two paths... We need to distinguish request and response cookies.
  //So either clear each time the cookies store before returning response
  //Or create a separate CookieParser for responses only and manually set 'Set-Cookie' header
  //But I decided to split CookieParser.cookies list into two parts
  // An original list of parsed request cookies.
  final List<Cookie> cookies = [];
  final List<Cookie> responseCookies = [];
  final String _secretKey;

  // creates new empty CookieParser, same as CookieParser.fromCookieValue(null)
  CookieParser([String secretKey = ""])
      : _secretKey =
            (secretKey + "00000000000000000000000000000000").substring(0, 32);

  /// Creates a new [CookieParser] by parsing the `Cookie` header [value].
  CookieParser.fromCookieValue(String? value, [String secretKey = ""])
      : this._secretKey =
            (secretKey + "00000000000000000000000000000000").substring(0, 32) {
    if (value != null) {
      cookies.addAll(_parseCookieString(value));
    }
  }

  /// Factory constructor to create a new instance from request [headers].
  factory CookieParser.fromHeader(Map<String, dynamic> headers,
      [String secretKey = ""]) {
    return CookieParser.fromCookieValue(
        headers[HttpHeaders.cookieHeader], secretKey);
  }

  /// Use this method to set Response's 'set-cookie' header
  /// Shelf now supports multiple 'set-cookie' values, that is why method returns List<String>
  List<String> toHeader() => responseCookies.map((c) => c.toString()).toList();

  /// Denotes whether the [cookies] list is empty.
  bool get isEmpty => cookies.isEmpty;
  bool get isResponseEmpty => responseCookies.isEmpty;

  /// Retrieves a cookie by [name].
  Cookie? get(String name) =>
      cookies.firstWhereOrNull((Cookie cookie) => cookie.name == name);

  /// Retrieves a response cookie by [name]. You will rarely need this, but in case.
  Cookie? getResponse(String name) =>
      responseCookies.firstWhereOrNull((Cookie cookie) => cookie.name == name);

  /// Adds a new cookie to [responseCookies] list.
  Cookie set(
    String name,
    String value, {
    String? domain,
    String? path,
    DateTime? expires,
    bool? httpOnly,
    bool? secure,
    int? maxAge,
  }) {
    var cookie = Cookie(name, value);
    if (domain != null) cookie.domain = domain;
    if (path != null) cookie.path = path;
    if (expires != null) cookie.expires = expires;
    if (httpOnly != null) cookie.httpOnly = httpOnly;
    if (secure != null) cookie.secure = secure;
    if (maxAge != null) cookie.maxAge = maxAge;

    // Update existing cookie, or append new one to list.
    var index = responseCookies.indexWhere((item) => item.name == name);
    if (index != -1) {
      responseCookies.replaceRange(index, index + 1, [cookie]);
    } else {
      responseCookies.add(cookie);
    }
    return cookie;
  }

  /// Retrieves a deciphered cookie by [name].
  // secretKey length must be exactly 32 bytes
  Future<Cookie?> getEncrypted(String name) async {
    final keyBytes = utf8.encode(_secretKey);
    if (keyBytes.length != 32)
      throw Exception(
          'Expected secretKey length is 32, but got: ${keyBytes.length}');
    final scookie = get(name);
    if (scookie == null) return null;
    //copy cookie to prevent raw data leak :D
    final cookie = Cookie(scookie.name, scookie.value)
      ..domain = scookie.domain
      ..expires = scookie.expires
      ..httpOnly = scookie.httpOnly
      ..maxAge = scookie.maxAge
      ..path = scookie.path
      ..secure = scookie.secure;
    var decoded = base64Url.decode(cookie.value);
    if (decoded.length <= 12 + 16)
      throw Exception('Wrong encrypted cookie length');
    final algorithm = AesGcm.with256bits();
    final key = await algorithm.newSecretKeyFromBytes(keyBytes);
    //can't use fromConcatenation constructor because of this:
    //https://github.com/dint-dev/cryptography/issues/55
    //final box = SecretBox.fromConcatenation(decoded, nonceLength: 12, macLength: 16);
    final cipherText = decoded.skip(12).take(decoded.length - 12 - 16).toList();
    final nonce = decoded.take(12).toList();
    final mac =
        decoded.skip(nonce.length + cipherText.length).take(16).toList();
    final box = SecretBox(cipherText, nonce: nonce, mac: Mac(mac));
    final bytes = await algorithm.decrypt(box, secretKey: key);
    cookie.value = utf8.decode(bytes);
    return cookie;
  }

  /// Adds a new ciphered cookie to [cookies] list.
  // secretKey length must be exactly 32 bytes
  Future<Cookie> setEncrypted(
    String name,
    String value, {
    String? domain,
    String? path,
    DateTime? expires,
    bool? httpOnly,
    bool? secure,
    int? maxAge,
  }) async {
    final keyBytes = utf8.encode(_secretKey);
    if (keyBytes.length != 32)
      throw Exception(
          'Expected secretKey length is 32, but got: ${keyBytes.length}');
    final valueBytes = utf8.encode(value);
    final algorithm = AesGcm.with256bits(nonceLength: 12);

    final key = await algorithm.newSecretKeyFromBytes(keyBytes);
    // Encrypt
    final secretBox = await algorithm.encrypt(valueBytes, secretKey: key);
    var encryptedValue = base64Url.encode(secretBox.concatenation());

    return set(name, encryptedValue,
        domain: domain,
        path: path,
        expires: expires,
        httpOnly: httpOnly,
        secure: secure,
        maxAge: maxAge);
  }

  /// Removes a cookie from list by [name].
  /// Note: this does not delete a cookie in the browser
  /// To remove the cookie from browser cache you need to set a new one
  /// with exactly same name, path and domain, but with the expires date in the past, value is irrelevant
  void remove(String name) =>
      responseCookies.removeWhere((Cookie cookie) => cookie.name == name);

  /// Clears the cookie list.
  void clear() => responseCookies.clear();
}

/// Parse a Cookie header value according to the rules in RFC 6265.
/// This function was adapted from `dart:io`.
List<Cookie> _parseCookieString(String s) {
  var cookies = <Cookie>[];

  int index = 0;

  bool done() => index == -1 || index == s.length;

  void skipWS() {
    while (!done()) {
      if (s[index] != " " && s[index] != "\t") return;
      index++;
    }
  }

  String parseName() {
    int start = index;
    while (!done()) {
      if (s[index] == " " || s[index] == "\t" || s[index] == "=") break;
      index++;
    }
    return s.substring(start, index);
  }

  String parseValue() {
    int start = index;
    while (!done()) {
      if (s[index] == " " || s[index] == "\t" || s[index] == ";") break;
      index++;
    }
    return s.substring(start, index);
  }

  bool expect(String expected) {
    if (done()) return false;
    if (s[index] != expected) return false;
    index++;
    return true;
  }

  while (!done()) {
    skipWS();
    if (done()) continue;
    String name = parseName();
    skipWS();
    if (!expect("=")) {
      index = s.indexOf(';', index);
      continue;
    }
    skipWS();
    String value = parseValue();
    try {
      cookies.add(Cookie(name, value));
    } catch (_) {
      // Skip it, invalid cookie data.
    }
    skipWS();
    if (done()) continue;
    if (!expect(";")) {
      index = s.indexOf(';', index);
      continue;
    }
  }

  return cookies;
}


example/example.dart
import 'dart:io';

import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf/shelf_io.dart' as io;
import 'package:shelf_cookie/shelf_secure_cookie.dart';

void main() {
  /// Request contains cookie header.
  /// e.g. 'cookie': 'ping=foo'

  /// use your 32-byte secure key during middleware construction
  /// don't save it in your source code btw
  /// consider something like this:
  /// final secretKey = myConfig.load('config.json').secretKey
  /// and add config.json to .gitignore
  final secretKey = "12345678901234567890123456789012";
  var handler = const shelf.Pipeline()
      .addMiddleware(cookieParser(secretKey))
      .addHandler((req) async {
    CookieParser cookies = req.context['cookies'] as CookieParser;
    if (cookies.get('ping') != null) {
      //secure: true - means send it via https only
      cookies.clear(); //or cookies = CookieParser(); to create a new instance
      cookies.setEncrypted('pong', 'bar', secure: true, httpOnly: true);
    }

    // Response will set cookie header.
    // e.g. 'set-cookie': 'pong=someencryptedandsignedvalue; Secure; HttpOnly'
    return shelf.Response.ok('check your cookies');
  });

  io.serve(handler, 'localhost', 8080).then((server) {
    print('Serving at http://${server.address.host}:${server.port}');
  });
}
